### DISCUSSION TITLE
Make axios usage in sdk tree shakeable?

### DISCUSSION THREAD

**Discussion by Jakob-em** [directus/directus#14293] [Feature Requests]
It would be great if the usage of axios in the sdk could be removed by tree shaking when a custom transport is used.
I am not sure if and how this is possible, but I would be happy to have a look at it if this is something that is desired.

Short background:
> I am currently working on an angular wrapper for the sdk and thought it might be cool if the angular HttpClient is used as transport and axios is removed from the client bundle.

**Comment by CapitaineToinon** [comment]
I second this. I believe however that the only way to achieve this would be to force users to specify what Transport class they want to use, even when they want to use axios. That would be a breaking change but a good one imo. Would also be very easy to implement. 

For example
```js
// when user wants to use the default transport
import { Transport, Directus } from '@directus/sdk';
const directus = new Directus('http://directus.example.com', { transport: new Transport() });

// or when the user wants to use their own implementation
import { ITransport, Directus } from '@directus/sdk';

class MyTransport extends ITransport {
  // ...
}

const directus = new Directus('http://directus.example.com', { transport: new MyTransport() });
```

I believe only this bit of code would need to be changed.
https://github.com/directus/directus/blob/main/packages/sdk/src/base/directus.ts#L87

I guess the same could be done for IStorage and IAuth?

**Reply by rijkvanzanten** [reply]
I'm a tad worried if the benefit is worth the tradeoff of the added complexity of trying to support multiple http abstractions. What would be the main benefit of relying on `HttpClient`? Just the size difference of not having to include `axios`? If that's the case, I'd much prefer dropping the abstraction complexity from the SDK in favor of using the browser native `fetch` or `XMLHttpRequest` instead, as that should work regardless of framework used and has the least added library code 🙂

**Reply by Jakob-em** [reply]
> I'm a tad worried if the benefit is worth the tradeoff of the added complexity of trying to support multiple http abstractions. What would be the main benefit of relying on `HttpClient`? Just the size difference of not having to include `axios`? If that's the case, I'd much prefer dropping the abstraction complexity from the SDK in favor of using the browser native `fetch` or `XMLHttpRequest` instead, as that should work regardless of framework used and has the least added library code 🙂 

Relying on the HttpClient instead of axios would add a few benefits besides the obvious (but small) bundle size difference.
The additional benefits are global interceptors for requests/responses, automatic polyfills for older browsers or the possibility to mock it in tests with angulars built in solutions (HttpClientTestingModule).
But I agree that the additional configuration complexity for the default usage is not a great solution.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
